relay/writer/file_util.go

```diff
@@ -0,0 +1,262 @@
+// It is not safe if there other routine is writing the file.
+func checkBinlogHeaderExist(filename string) (bool, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "open file %s", filename)
+	}
+	defer f.Close()
+
+	return checkBinlogHeaderExistFd(f)
+}
+
+// checkBinlogHeaderExistFd checks if the file has a binlog file header.
+// It is not safe if there other routine is writing the file.
+func checkBinlogHeaderExistFd(fd *os.File) (bool, error) {
+	fileHeaderLen := len(replication.BinLogFileHeader)
+	buff := make([]byte, fileHeaderLen)
+	n, err := fd.Read(buff)
+	if err != nil {
+		if n == 0 && err == io.EOF {
+			return false, nil // empty file
+		}
+		return false, errors.Annotate(err, "read binlog header")
+	} else if n != fileHeaderLen {
+		return false, errors.Errorf("binlog file %s has no enough data, only got % X", fd.Name(), buff[:n])
+	}
+
+	if !bytes.Equal(buff, replication.BinLogFileHeader) {
+		return false, errors.Errorf("binlog file %s header not valid, got % X, expect % X", fd.Name(), buff, replication.BinLogFileHeader)
+	}
+	return true, nil
+}
+
+// checkFormatDescriptionEventExist checks if the file has a valid FormatDescriptionEvent.
+// It is not safe if there other routine is writing the file.
+func checkFormatDescriptionEventExist(filename string) (bool, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return false, errors.Annotatef(err, "open file %s", filename)
+	}
+	defer f.Close()
+
+	// FormatDescriptionEvent always follows the binlog file header
+	exist, err := checkBinlogHeaderExistFd(f)
+	if err != nil {
+		return false, errors.Annotatef(err, "check binlog file header for %s", filename)
+	} else if !exist {
+		return false, errors.Errorf("no binlog file header at the beginning for %s", filename)
+	}
+
+	// check whether only the file header
+	fileHeaderLen := len(replication.BinLogFileHeader)
+	fs, err := f.Stat()
+	if err != nil {
+		return false, errors.Errorf("get stat for %s", filename)
+	} else if fs.Size() == int64(fileHeaderLen) {
+		return false, nil // only the file header
+	}
+
+	// seek to the beginning of the FormatDescriptionEvent
+	_, err = f.Seek(int64(fileHeaderLen), io.SeekStart)
+	if err != nil {
+		return false, errors.Annotatef(err, "seek to %d for %s", fileHeaderLen, filename)
+	}
+
+	// parse a FormatDescriptionEvent
+	var found bool
+	onEventFunc := func(e *replication.BinlogEvent) error {
+		if e.Header.EventType != replication.FORMAT_DESCRIPTION_EVENT {
+			return errors.Errorf("got %+v, expect FormatDescriptionEvent", e.Header)
+		} else if (e.Header.LogPos - e.Header.EventSize) != uint32(fileHeaderLen) {
+			return errors.Errorf("wrong offset %d for FormatDescriptionEvent, should be %d", e.Header.LogPos, fileHeaderLen)
+		}
+		found = true
+		return nil
+	}
+
+	// only parse single event
+	eof, err := replication.NewBinlogParser().ParseSingleEvent(f, onEventFunc)
+	if err != nil {
```

Comment:
if some events after `format description event` are damaged, but `found == true`, should we return true?

Reply:
I changed to return `true` even `err != nil` in https://github.com/pingcap/dm/pull/117/commits/6d54140244c2bbf5adc5087948505703e81f14a3, because this function only checks if `FormatDescriptionEvent` exists.

Reply:
make sense